### PR TITLE
auditdb: extended transaction tracking #242

### PR DESCRIPTION
**Pull request by adecaro**
We have extended AuditDB to track Transactions as requested by the associated git issue.

Signed-off-by: Angelo De Caro <adc@zurich.ibm.com>

### CODE REVIEW

**Review by mbrandenburger — COMMENTED** [review]
Thanks Angelo!

**Comment by adecaro** [comment]
Hi @mbrandenburger , thanks a lot for your review. I have addressed your comments :)

**Review by mbrandenburger — COMMENTED** [review]
Thanks Angelo! Overall this looks good to me - just a few minors :)

File: token/sdk/sdk.go [hunk 1]
```diff
@@ -58,9 +56,6 @@ func (p *SDK) Install() error {
 	}
 	logger.Infof("Token platform enabled, installing...")
 
-	// event service
-	assert.NoError(p.registry.RegisterService(&events.Service{EventSystem: simple.NewEventBus()}))
-
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/sdk/sdk.go:63]: Yes :)

File: token/services/auditor/auditdb/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+```go
+	auditRecord, err := tx.TokenRequest.AuditRecord()
+	if err != nil {
+		return errors.WithMessagef(err, "failed getting audit records for tx [%s]", tx.ID())
+	}
+	if err := auditDB.Append(auditRecord); err != nil {
+		return errors.WithMessagef(err, "failed appening audit records for tx [%s]", tx.ID())
+	}
+```
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/services/auditor/auditdb/README.md:33]: what's wrong here, sorry? 😓

File: token/services/auditor/auditdb/db/badger/badger.go [hunk 3]
```diff
@@ -222,20 +309,170 @@ func (db *Persistence) Query(ids []string, types []string, status []driver.Statu
...
+		found := false
+		for _, st := range m.txStatuses {
+			if record.Record.Status == st {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/services/auditor/auditdb/db/badger/badger.go:465]: This comment is on outdated code. I have already introduced a filter function there 🤔

File: token/services/network/network.go [hunk 4]
```diff
@@ -291,29 +310,46 @@ func (n *Network) ComputeTxID(id *TxID) string {
...
 	return &LocalMembership{lm: n.n.LocalMembership()}
 }
 
+// GetEnrollmentID returns the enrollment ID bound to the passed marshalled audit info
 func (n *Network) GetEnrollmentID(raw []byte) (string, error) {
 	return n.n.GetEnrollmentID(raw)
 }
 
+// SubscribeTxStatusChanges register a listener for transaction status updates for the given id.
+// TODO: introduce a TTL
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/services/network/network.go:334]: Agreed. Removing it.

File: token/services/ttx/auditor.go [hunk 5]
```diff
@@ -166,30 +169,18 @@ func NewAuditApproveView(w *token.AuditorWallet, tx *Transaction) *AuditApproveV
...
+	logger.Debugf("store audit records...done")
 
-	if logger.IsEnabledFor(zapcore.DebugLevel) {
-		logger.Debugf("sign and send back")
-	}
+	logger.Debugf("sign and send back")
 	if err := a.signAndSendBack(context); err != nil {
 		return nil, err
 	}
+	logger.Debugf("audit approve done")
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/services/ttx/auditor.go:182]: Agreed.

File: token/services/auditor/auditdb/auditdb.go [hunk 6]
```diff
@@ -274,14 +513,26 @@ func (cm *Manager) AuditDB(w *token.AuditorWallet) (*AuditDB, error) {
...
+
+// GetAuditDB returns the AuditDB for the given auditor wallet.
+// Nil might be returned if the auditor wallet is not found or an error occurred.
 func GetAuditDB(sp view2.ServiceProvider, w *token.AuditorWallet) *AuditDB {
-	s, err := sp.GetService(&Manager{})
+	if w == nil {
+		logger.Debugf("no auditor wallet provided")
+		return nil
+	}
+	s, err := sp.GetService(managerKey)
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/services/auditor/auditdb/auditdb.go:527]: Agreed.

File: token/services/auditor/auditdb/README.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+   auditDB := auditdb.GetAuditDB(context, auditorWallet)
+```
+
+## Append Audit Record
+
+An `Audit Record` can be obtained from a `Token Request`. 
+Usually Token Requests are themselves embedded in token transactions.
+
+Here is an example of extraction of an audit record from a token request, and
+appending of the record to the audit DB:
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/services/auditor/auditdb/README.md:23]: Agreed.

File: token/services/auditor/auditdb/db/badger/badger.go [hunk 8]
```diff
@@ -39,11 +43,11 @@ type TransactionRecord struct {
 }
 
 type Persistence struct {
-	db *badger.DB
-
-	seq     *badger.Sequence
-	txn     *badger.Txn
-	txnLock sync.Mutex
+	db          *badger.DB
+	numGoStream int
```
Comment (mbrandenburger) [inline token/services/auditor/auditdb/db/badger/badger.go:47]: Plural? Steams?

**Review by mbrandenburger — APPROVED** [review]
LGTM!

File: token/services/auditor/auditdb/db/badger/badger.go (hunk 8 above)
Comment (adecaro) [reply token/services/auditor/auditdb/db/badger/badger.go:47]: singular because it is number of go routines per stream 😅

**Pull request merged**